proposed/2022/duplicate-nuget-item-error-handling.md

```diff
@@ -0,0 +1,142 @@
+
+1. SDK + MSBuild.exe/dotnet.exe
+
+NuGet completely controls the reading of the items. By adding the warning in the collector targets, the warnings can be raised.
+When the warning is elevated to an error, the error scenario is fail fast, so restore *will not* run. This is unusual and different from the other input errors.
+
+1. SDK + Visual studio
+
+The project-system calls the `Collect{ItemName}s` targets. These warnings will be raised in this target.
+For design time purposes, we respect the `ContinueOnError` property.
+When the warning is elevated to an error, no items for that target are nominated.
+This can lead to an awkward scenario where many errors appear in the user's error list.
+
+It is possible that restore succeeds in this case, despite the collect target erroring out. This is another tecnhical limitation.
+
+1. SDK + NuGet.exe
+
+NuGet.exe shells out to MSBuild to generate the restore graph. The warnings from the targets are swallowed in this scenario. Due to the support for TreatWarningsAsErrors, to prevent silent failures, NuGet.exe restore will not raise any warnings/errors.
+
+1. Legacy csproj + MSBuild.exe
+
+NuGet completely controls the reading of the items. By adding the warning in the collector targets, the warnings can be raised.
+When the warning is elevated to an error, the error scenario is fail fast, so restore *will not* run. This is unusual and different from the other input errors.
+
+1. Legacy csproj + Visual studio
+
+The legacy project system does not call the `Collect{ItemName}s` targets. As such, raising a warning without an API change is difficult. There are arguably better designs that'd work for both legacy & CPS based PackageReference but the return of investment is not super high.
+
+1. Legacy csproj + NuGet.exe
+
+NuGet.exe shells out to MSBuild to generate the restore graph. The warnings from the targets are swallowed in this scenario. Due to the support for TreatWarningsAsErrors, to prevent silent failures, NuGet.exe restore will not raise any warnings/errors.
+
+## Drawbacks
+
+This is an imperfect, best effort solution. There are *many* caveats.
+
+- NuGet.exe will not be able to raise warnings.
+- Legacy csproj projects will not raise warnings/errors in Visual Studio.
+- SDK based projects in Visual Studio do not nominate that item when the warning has been elevated to an error, leading to a potentially challenging experience.
+- SDK based projects in Visual Studio that have errors due to duplicate items may succeed at restore or even build time, thus creating a false positive build.
+
+## Rationale and alternatives
+
+- Deduplicate silently. We can take a similar approach and deduplicate silently. This will ensure consistent behavior, but it may cause some unexpected behavior.
+
+## Prior Art
+
+- SDK does [duplicate checking](https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateItems.cs).
+
+## Unresolved Questions
+
```

Comment:
In MSbuild structured log viewer, there's a concept of Reassigned properties:

![image](https://user-images.githubusercontent.com/1192347/157769672-2e03a191-8e55-4088-8a14-4efe2779f195.png)

Can we say that duplicate PackageReference's are 'Item reassignments'?

Reply:
Not really. The same concept doesn't exist for items.